Change signature of File::try_lock and File::try_lock_shared

cberner opened the pull request:
These methods now return Result<(), TryLockError> instead of Result<bool, Error> to make their use less errorprone

These methods are unstable under the "file_lock" feature. The related tracking issue is https://github.com/rust-lang/rust/pull/130994 and this PR changes the signatures as discussed by libs-api: https://github.com/rust-lang/rust/issues/130994#issuecomment-2770838848

cberner committed:
Change signature of File::try_lock and File::try_lock_shared

These methods now return Result<(), TryLockError> instead of
Result<bool, Error> to make their use less errorprone

juntyr commented on library/std/src/fs.rs:
```diff
@@ -116,6 +116,19 @@ pub struct File {
+/// [`try_lock`]: File::try_lock
+/// [`try_lock_shared`]: File::try_lock_shared
+#[unstable(feature = "file_lock", issue = "130994")]
+pub enum TryLockError {
```
Could the error type also implement core::error::Error?

cberner replied:
yep! I just added that

Noratrieb replied:
I can't find it in this file.

cberner replied:
Oh sorry, I reverted it due to the discussion in the tracking issue about omitting the From implementation and just sticking to the signature change. Want me to add the Error impl back?

eric-seppanen commented on library/std/src/fs.rs:
```diff
@@ -116,6 +116,19 @@ pub struct File {
+/// [`try_lock`]: File::try_lock
+/// [`try_lock_shared`]: File::try_lock_shared
+#[unstable(feature = "file_lock", issue = "130994")]
+pub enum TryLockError {
+    /// The lock could not be acquired due to an I/O error on the file.
+    Error(io::Error),
+    /// The lock could not be acquired at this time because the operation would block.
+    WouldBlock,
+}
```
Should this document that `TryLockError::Error` containing `ErrorKind::WouldBlock` will never be returned by std?

Noratrieb replied:
I think that makes sense, yes

cberner committed:
Implement error::Error for TryLockError

cberner commented:
Ok, I pushed a new commit to address the comments and also made the docs a little more clear for `TryLockError::WouldBlock`

cberner commented:
@Noratrieb anything I can do to help move this PR forward?

jieyouxu commented:
r? libs

workingjubilee commented:
Seems fine to me. People can continue to kibitz over what it should be if they want. @bors r+ rollup

bors commented:
:pushpin: Commit 7302f8eecd1bb3502c7189514ff00990af0988fc has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee 7302f8eecd1bb3502c7189514ff00990af0988fc -->
<!-- homu: {"type":"Approved","sha":"7302f8eecd1bb3502c7189514ff00990af0988fc","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

GuillaumeGomez commented:
Failed in https://github.com/rust-lang/rust/pull/140562:

```
unused import: `TryLockError`
library/std/src/fs/tests.rs:4:53
```

@bors r-

workingjubilee commented:
welp.

cberner commented:
@bors ready

cberner commented:
Oops, my bad. I missed the `#[cfg]` for that import. I fixed it now, and also added one to the check_would_block macro definition. @workingjubilee can I trouble you for another stamp?

cberner commented:
@rustbot ready

workingjubilee commented on library/std/src/fs/tests.rs:
```diff
@@ -28,6 +36,23 @@ macro_rules! check {
+macro_rules! check_would_block {
+    ($e:expr) => {
+        match $e {
+            Ok(_) => panic!("{} acquired lock when it should have failed", stringify!($e)),
+            Err(TryLockError::WouldBlock) => (),
+            Err(e) => panic!("{} failed with: {e}", stringify!($e)),
```
wait.

why isn't this just [`assert_matches!`](https://doc.rust-lang.org/std/assert_matches/macro.assert_matches.html)?

cberner replied:
No particular reason. I was just following the style of the `check` macro above. And it did seem nice to have clear error messages to differentiate why `Ok(_)` is a failure vs an IO error.

I don't have a strong preference though. Want me to remove the `check_would_block` macro and use `assert_matches` instead?

workingjubilee replied:
Ah... well, that macro dates back to before Rust 1.0. It has never been updated stylistically despite evolutions in the language which can make code like that somewhat easier to write and read.

cberner replied:
Ah, got it. I didn't realize it was that old! Ok, updated PR coming :)

workingjubilee replied:
This was the where/when it was added: https://github.com/rust-lang/rust/blob/6bfbad937bdf578e35777d079f8dcfab49758041/src/libstd/fs.rs

cberner replied:
Oh, wow :sweat_smile: 

Ok, I've updated it, and Github CI is green. Hopefully I didn't miss anything else for bors' test suite.

Thanks for your patience with this!

workingjubilee commented:
Thanks! Let's try this again.

@bors r+ rollup
